File: debt-analyser/tools/debt-evolution.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strconv"
	"strings"

	"github.com/tealeg/xlsx"
)

// compareFiles compares two .xlsx files and returns the count of added and removed strings.
func compareFiles(file1, file2 string) (int, int, map[string]int, map[string]int) {
	wb1, err := xlsx.OpenFile(file1)
	if err != nil {
		panic(err) // Simplified error handling
	}
	wb2, err := xlsx.OpenFile(file2)
	if err != nil {
		panic(err)
	}

	// Assuming data is in the first sheet and in the first column
	ws1 := wb1.Sheets[0]
	ws2 := wb2.Sheets[0]

	count1 := make(map[string]int)
	count2 := make(map[string]int)

	for _, row := range ws1.Rows {
		if len(row.Cells) > 0 {
			val := row.Cells[0].String()
			if val != "" {
				count1[val]++
			}
		}
	}

	for _, row := range ws2.Rows {
		if len(row.Cells) > 0 {
			val := row.Cells[0].String()
			if val != "" {
				count2[val]++
			}
		}
	}

	added := 0
	removed := 0

	for key, val := range count2 {
		if _, exists := count1[key]; !exists {
			added += val
		} else {
			diff := val - count1[key]
			if diff > 0 {
				added += diff
			}
		}
	}

	for key, val := range count1 {
		if _, exists := count2[key]; !exists {
			removed += val
		} else {
			diff := val - count2[key]
			if diff > 0 {
				removed += diff
			}
		}
	}

	return added, removed, count1, count2
}

// processDirectory processes each directory and its files.
func processDirectory(rootDir string) {
	remDict := make(map[string][2]int)
	var dirs []string

	filepath.Walk(rootDir, func(path string, info os.FileInfo, err error) error {
		if err != nil {
			return err
		}
		if info.IsDir() && path != rootDir {
			dirs = append(dirs, path)
		}
		return nil
	})

	for _, dir := range dirs {
		files, err := filepath.Glob(filepath.Join(dir, "*.xlsx"))
		if err != nil {
			panic(err) // Simplified error handling
		}
		sort.Slice(files, func(i, j int) bool {
			numI, _ := strconv.Atoi(strings.Split(filepath.Base(files[i]), ".")[0])
			numJ, _ := strconv.Atoi(strings.Split(filepath.Base(files[j]), ".")[0])
			return numI < numJ
		})

		for i := 0; i < len(files)-1; i++ {
			file1 := files[i]
			file2 := files[i+1]
			added, removed, count1, _ := compareFiles(file1, file2)
			if i == 0 {
				remDict[dir] = [2]int{len(count1), 0}
			}
			current := remDict[dir]
			remDict[dir] = [2]int{current[0] + added, current[1] + removed}
		}
	}

	removedInstances := 0
	for _, v := range remDict {
		if v[1] > 0 {
			removedInstances++
		}
	}

	fmt.Printf("Occurrence of debt removal: %f%%\n", float64(removedInstances)*100/float64(len(remDict)))
}

func main() {
	rootDir := "../debt_data/contracts"
	processDirectory(rootDir)
}

File: debt-analyser/tools/debt-evolution.py
import os
from collections import Counter
from openpyxl import load_workbook


def compare_files(file1, file2):
    """Compare two files and return the count of added and removed strings from .xlsx files."""
    wb1 = load_workbook(filename=file1)
    wb2 = load_workbook(filename=file2)
    
    # Assuming data is in the first sheet and in the first column
    ws1 = wb1.active
    ws2 = wb2.active
    
    # Extract strings from the first column
    strings1 = [row[0] for row in ws1.iter_rows(min_row=1, values_only=True) if row[0] is not None]
    strings2 = [row[0] for row in ws2.iter_rows(min_row=1, values_only=True) if row[0] is not None]
    
    # Use Counter to count occurrences of each string
    count1 = Counter(strings1)
    count2 = Counter(strings2)
    
    # Find differences
    added = sum((count2 - count1).values())
    removed = sum((count1 - count2).values())
    
    return added, removed, count1, count2


def process_directory(root_dir):
    """Process each directory and its files."""
    rem_dict = dict()
    for root, dirs, files in os.walk(root_dir):
        if len(dirs) == 0:  # This means we are in a leaf directory
            sorted_files = sorted(files, key=lambda x: int(x.split('.')[0]))  # Sort files by their numerical name
            for i in range(len(sorted_files) - 1):
                file1 = os.path.join(root, sorted_files[i])
                file2 = os.path.join(root, sorted_files[i + 1])
                added, removed, count1, _ = compare_files(file1, file2)
                if i == 0:
                    rem_dict[root] = [sum(count1.values()), 0]
                rem_dict[root] = [rem_dict[root][0] + added, rem_dict[root][1] + removed] 
                # print(f"Between {sorted_files[i]} and {sorted_files[i + 1]}: {added} added, {removed} removed in {root}")
    
    removed_instances = 0
    for k, v in rem_dict:
        if v[1] > 0:
            removed_instances += 1
    
    print(f'Occurence of debt removal: ${removed_instances * 100 / len(rem_dict.keys())}')


root_dir = '../debt_data/contracts'
process_directory(root_dir)